### PR TITLE
Engine - Enhanced Syntax for Windows Decoder

### PR DESCRIPTION
**Pull request by soynof**
|Related issue|
|---|
|#18458|

## Description:
In light of the latest advancements in the engine, we are undertaking changes to refine the syntax used in our Windows decoders. These updates are aimed at enhancing readability and ensuring user-friendliness.

Key objectives for this update include:

- [x] Enhanced Event.Original Handling: We will be revisiting the handling of event.original within the system decoders. Going forward, this field will consistently reflect the original event as captured by the agent, as opposed to its representation after being sent by the agent.

- [x] Check Stages: Our team will be working to update the check stages within the decoders. This entails optimizing the process to ensure smoother and more efficient checks for various conditions.

- [x] Syntax and Visualization Improvements: During this process, we will meticulously review the arrangement of stages, spacing between expressions, and field names. The aim is to create a more intuitive and comprehensible visual structure for users, making it easier to interpret the decoding process.

This PR also includes code modifications:
- Checks syntax is adjusted, now values in an expression type <$REF> <OP> <VALUE>, supports VALUE to be a text between single quotes that will be translated as a json string.
- Fixed a bug in the map. When we assign a map from a decoder, the keys can be numeric without being translated to an array.
- Added functionality to json::Json
- Helpers now support single-quoted arguments, they can have literal text enclosed in single quotes and support empty arguments like `''`.

### CODE REVIEW

**Review by juliancnn — APPROVED** [review]
LGTM :eyes:

File: src/engine/source/builder/helperParser.hpp [hunk 1]
```diff
@@ -208,6 +213,22 @@ inline parsec::Parser<HelperToken> getHelperParser(bool eraseScapeChars = false)
         return parsec::makeSuccess(std::string(1, syntax::PARENTHESIS_OPEN), next);
     };
 
+    parsec::Parser<std::string> parenthCloseParser = [](auto sv, auto pos) -> parsec::Result<std::string>
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:216]: nitpick: use auto

File: src/engine/source/builder/helperParser.hpp [hunk 2]
```diff
@@ -218,7 +239,17 @@ inline parsec::Parser<HelperToken> getHelperParser(bool eraseScapeChars = false)
         return parsec::makeSuccess(std::string(1, syntax::PARENTHESIS_CLOSE), pos);
     };
 
-    parsec::Parser<std::string> argParser = [eraseScapeChars](auto sv, auto pos) -> parsec::Result<std::string>
+    parsec::Parser<std::string> behindParenthOpenParser = [](auto sv, auto pos) -> parsec::Result<std::string>
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:242]: ditto

File: src/engine/source/builder/helperParser.hpp [hunk 3]
```diff
@@ -274,6 +306,66 @@ inline parsec::Parser<HelperToken> getHelperParser(bool eraseScapeChars = false)
...
+        if (!valid)
+        {
+            return parsec::makeError<std::string>("Invalid single quote string", pos);
+        }
+
+        // TODO: Add test for both cases
+        std::string arg {};
+        if (eraseScapeChars)
+        {
+            if (escapedFirstReference){
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:344]: nitpick: apply clang format

File: src/engine/source/builder/helperParser.hpp [hunk 4]
```diff
@@ -274,6 +306,66 @@ inline parsec::Parser<HelperToken> getHelperParser(bool eraseScapeChars = false)
         return parsec::makeSuccess(std::move(arg), next);
     };
 
+    parsec::Parser<std::string> quotedArgParser = [eraseScapeChars](auto sv, auto pos) -> parsec::Result<std::string>
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:309]: ditto

File: src/engine/test/source/builder/builders/operationBuilder_test.cpp [hunk 5]
```diff
@@ -850,9 +852,10 @@ TEST(OperationMapBuilderTest, BuildsOperatesObject)
     auto expression = getOperationMapBuilder(helperRegistry, schemf::mocks::EmptySchema::create())(
         definition, std::make_shared<defs::mocks::FailDef>());
     auto expressionsRootLevel = expression->getPtr<base::Operation>()->getOperands();
-    auto expressionsNestedLevel = expressionsRootLevel[6]->getPtr<base::Operation>()->getOperands();
+    // 6 is the number of mapping in the root level and 1 is the created object if not exists
+    auto expressionsNestedLevel = expressionsRootLevel[7+1]->getPtr<base::Operation>()->getOperands();
```
Comment (NahuFigueroa97) [inline src/engine/test/source/builder/builders/operationBuilder_test.cpp:856]: magic numbers are not accepted, please define this as a constexpr.

File: src/engine/source/base/utils/stringUtils.hpp [hunk 6]
```diff
@@ -103,6 +103,28 @@ inline bool startsWith(std::string_view str, std::string_view prefix)
...
+ * @param escapedChars A string containing all characters that can be escaped.
+ * @param strictMode If true an exception is thrown if an invalid escape sequence is found.
+ * @return The unescaped string.
+ * @todo: Add support for strict mode where an exception is thrown if an invalid escape sequence is found.
+ */
+std::string
+unescapeString(std::string_view str, char escapeChar, const std::string& escapedChars, const bool strictMode = false);
+
+inline std::string
+unescapeString(std::string_view str, char escapeChar, const char escapedChar, const bool strictMode = false) {
```
Comment (NahuFigueroa97) [inline src/engine/source/base/utils/stringUtils.hpp:124]: nitpick: apply clang format

File: src/engine/source/builder/helperParser.hpp [hunk 7]
```diff
@@ -53,7 +53,12 @@ inline std::tuple<std::string, json::Json> toBuilderInput(const HelperToken& hel
                                                           const std::string& targetField)
 {
     json::Json value {};
-    auto valueStr = fmt::format("{}({})", helperToken.name, base::utils::string::join(helperToken.args, ","));
+
+    std::string strArgs = helperToken.args.size() == 1 && helperToken.args[0].empty()
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:57]: nitpick: use auto

File: src/engine/source/builder/helperParser.hpp [hunk 8]
```diff
@@ -218,7 +239,17 @@ inline parsec::Parser<HelperToken> getHelperParser(bool eraseScapeChars = false)
...
+    {
+        if (pos == 0 || sv[pos - 1] != syntax::PARENTHESIS_OPEN)
+        {
+            return parsec::makeError<std::string>("Parenthesis open expected", pos);
+        }
+
+        return parsec::makeSuccess(std::string(1, syntax::PARENTHESIS_OPEN), pos);
+    };
+
+    parsec::Parser<std::string> simpleArgParser = [eraseScapeChars](auto sv, auto pos) -> parsec::Result<std::string>
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:252]: ditto

File: src/engine/source/builder/helperParser.hpp [hunk 9]
```diff
@@ -402,7 +503,43 @@ inline parsec::Parser<ExpressionToken> getExpressionParser()
         return parsec::makeSuccess(std::move(word), next);
     };
 
-    auto valueParser = valueRefParser | jsonParser | wordParser;
+    parsec::Parser<json::Json> singleQuotesParser = [](auto sv, auto pos) -> parsec::Result<json::Json>
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:506]: ditto

File: src/engine/source/builder/builders/operationBuilder.cpp [hunk 10]
```diff
@@ -137,6 +137,50 @@ enum class OperationType
...
+auto deleteEmptyObjectTerm(const std::string& field) -> base::Expression
+{
+    auto name {fmt::format("unmap.ifEmpty.value[{}]", field)};
+    auto successTrace {fmt::format("[{}] -> Success", name)};
+
+    auto fn = [field, successTrace](const auto& e)
+    {
+        if (e->isObject(field) && e->isEmpty(field))
+        {
+            e->erase(field);}
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/builders/operationBuilder.cpp:178]: apply clang format

File: src/engine/source/base/utils/stringUtils.cpp [hunk 11]
```diff
@@ -74,4 +74,36 @@ std::vector<std::string> splitEscaped(std::string_view input, const char& splitC
...
 }
 
+std::string
+unescapeString(std::string_view str, char escapeChar, const std::string& escapedChars, const bool strictMode)
+{
+    std::string result;
+    result.reserve(str.size());  // Reserve memory upfront to avoid reallocations
+
+    auto it = str.begin();
+    while (it != str.end())
```
Comment (NahuFigueroa97) [inline src/engine/source/base/utils/stringUtils.cpp:84]: nitpick: could you use std::accumulate of almorithm

File: src/engine/source/builder/helperParser.hpp [hunk 12]
```diff
@@ -402,7 +503,43 @@ inline parsec::Parser<ExpressionToken> getExpressionParser()
...
+
+        if (!valid)
+        {
+            return parsec::makeError<json::Json>("Invalid single quote string", pos);
+        }
+
+        const auto quoted = sv.substr(pos + 1, next - pos - 1);
+        const auto str = unescapeString(quoted, syntax::FUNCTION_HELPER_DEFAULT_ESCAPE, syntax::SINGLE_QUOTE, false);
+
+        json::Json value;
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/helperParser.hpp:536]: nitpick: initialize variable

File: src/engine/source/builder/builders/operationBuilder.cpp [hunk 13]
```diff
@@ -137,6 +137,50 @@ enum class OperationType
...
+    return base::Term<base::EngineOp>::create("setObjectOp", fn);
+}
+
+/**
+ * @brief Delete the Object if it is empty
+ *
+ * @param targetField
+ * @return base::Expression
+ */
+auto deleteEmptyObjectTerm(const std::string& field) -> base::Expression
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/builders/operationBuilder.cpp:169]: use string_view

File: src/engine/source/builder/builders/operationBuilder.cpp [hunk 14]
```diff
@@ -137,6 +137,50 @@ enum class OperationType
     FILTER
 };
 
+/**
+ * @brief Set the Object if it does not exist
+ *
+ * @param targetField
+ * @return base::Expression
+ */
+auto setObjectTerm(const std::string& field) -> base::Expression
```
Comment (NahuFigueroa97) [inline src/engine/source/builder/builders/operationBuilder.cpp:146]: use string_view

File: src/engine/source/builder/helperParser.hpp (hunk 12 above)
Comment (juliancnn) [reply src/engine/source/builder/helperParser.hpp:536]: It is not a native type, so it follows the initialization rules, using its default parameterless constructor:
https://en.cppreference.com/w/cpp/language/data_members#Member_initialization
Immediately after creating it with its default value (json null) it is replaced by a "json string"

File: src/engine/source/builder/helperParser.hpp (hunk 9 above)
Comment (juliancnn) [reply src/engine/source/builder/helperParser.hpp:507]: ditto

**Review by JavierBejMen — APPROVED** [review]
LGTM

**Review by NahuFigueroa97 — APPROVED** [review]
LGTM!

**Pull request merged**